---
title: "Exercise4"
author: "Asaf Pinhasi"
date: "2022-11-18"
output: html_document
---

```{r setup, include=FALSE, warning = F, message=F}
knitr::opts_chunk$set(echo = TRUE)
```

### Exploring the count tables

Here, import an example count table and do some exploration of the expression data. 

```{r exSetup1, warning = F, message=F}
# counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
#                            package = "compGenomRData")
# coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv", 
#                             package = "compGenomRData")

#loading dataframes from files

setwd("C:/Users/KerenYlab/Asaf/R Genomic Data Analysis/data")

coldata_df <- read.delim("SRP029880.colData.tsv", stringsAsFactors = F)
counts_df <- read.delim("SRP029880.raw_counts.tsv")

# importing relevant libraries
library(pheatmap)
library(gridExtra)
library(corrplot)
```

1. Normalize the counts using the TPM approach. 

```{r Q1}
#get lengths of genes from width column
counts <- as.matrix(counts_df)

geneLengths <- as.vector(subset(counts, select = c(width)))
#normalize by geneLength
rpk <- apply(subset(counts, select = c(-width)), 2,
             function(x) x/(geneLengths/1000))
#normalize by the sample size using normalized rpk values
tpm <- apply(rpk, 2, function(x) x / sum(as.numeric(x)) * 10^6)
```

2. Plot a heatmap of the top 500 most variable genes. Compare with the heatmap obtained using the 100 most variable genes. Hint : use var() function to compute the variance. 

```{r Q2}
variances <- apply(tpm, 1, var)
variances <- sort(variances, decreasing = T)
#select most variable genes from variances (sorted by variance)
selected_500_Genes <- names(variances[1:500])
selected_100_Genes <- names(variances[1:100])

#creates heatmaps for most variable genes
p1 <- pheatmap(tpm[selected_500_Genes,], scale = 'row', show_rownames = FALSE, main = "top 500 most variable")
p2 <- pheatmap(tpm[selected_100_Genes,], scale = 'row', show_rownames = FALSE, main = "top 100 most variable")
#plot the two heatmaps side by side
plot_list=list()
plot_list[['p1']]=p1[[4]]
plot_list[['p2']]=p2[[4]]
grid.arrange(grobs=plot_list, ncol=2)
```

3. Re-do the heatmaps setting the `scale` argument to `none`, and `column`. Compare the results with `scale = 'row'`.

```{r Q3p1}
#scales the heatmaps according to the values on the different axis
p1 <- pheatmap(tpm[selected_100_Genes,], scale = 'row', show_rownames = FALSE, main = "scale = row", fontsize_col = 6)
p2 <- pheatmap(tpm[selected_100_Genes,], scale = 'none', show_rownames = FALSE, main = "scale = none", fontsize_col = 6)
p3 <- pheatmap(tpm[selected_100_Genes,], scale = 'column', show_rownames = FALSE, main = "scale = column", fontsize_col = 6)
```
```{r Q3p2, out.width = "90%"}
# plot the 3 heatmaps side by side
plot_list=list()
plot_list[['p1']]=p1[[4]]
plot_list[['p2']]=p2[[4]]
plot_list[['p3']]=p3[[4]]
grid.arrange(grobs=plot_list, ncol=3)
```

4. Draw a correlation plot for the samples depicting the sample differences as 'ellipses', drawing only the upper end of the matrix, and order samples by hierarchical clustering results based on `average` linkage clustering method. 
```{r Q4}
correlationMatrix <- cor(tpm)


corrplot(correlationMatrix, method = 'ellipse', order = 'hclust',
         hclust.method = 'average', addrect = 2, type = 'upper')
```

### Git and Github

5. Open Github account, and install git if it required as instructed in the following link: https://github.com/llendway/github_for_collaboration/blob/master/github_for_collaboration.md

6. Create a repository and commit and push the answers of this exercise through Rtudio (Please upload the answers to the exercises also to Moodle).

7. Insert the link to your Github account here.
link:
https://github.com/asafpinhasi/Genomic-Data-Analysis.git